### PR TITLE
Semver parse fix to handle prereleases and build parts

### LINKED ISSUE #1324: SemVerHelper.parse doesn't handle prerelease segments that begin with a digit.
### Description

The SemVerHelper.parse method returns a None prerelease when given a SemVer string like "0.0.12-012345678".  This fails according to the spec, quoted here:

> A pre-release version MAY be denoted by appending a hyphen and a series of dot separated identifiers immediately following the patch version. Identifiers MUST comprise only ASCII alphanumerics and hyphen [0-9A-Za-z-]. Identifiers MUST NOT be empty. Numeric identifiers MUST NOT include leading zeroes. Pre-release versions have a lower precedence than the associated normal version. A pre-release version indicates that the version is unstable and might not satisfy the intended compatibility requirements as denoted by its associated normal version. Examples: 1.0.0-alpha, 1.0.0-alpha.1, 1.0.0-0.3.7, 1.0.0-x.7.z.92.

### Repro steps

Please provide the steps required to reproduce the problem

1. `Fake.SemVerHelper.parse "0.0.12-01234578" |> fun i -> i.Prerelease.IsNone <> true`

### Expected behavior

The prerelease segment is parsed as Some { Origin = "012345678"; Name = ""; Number = Some "012345678" }.

### Actual behavior

The prerelease segment is parsed as None

### Known workarounds

None

### PR DESCRIPTION
**Pull request by baronfel**
Fixes #1324 by rewriting the prerelease and build parsing sections to adhere to spec, while maintaining the Origin, Name, and Number portions of the current logic.

This has implications for existing uses of this code, because the prior code wasn't up to spec.  This code implements proper comparison for prerelease identifiers, and now discards the build portion in order to compare.

Thoughts?  We hit this when a git sha that we used for stamping prerelease information on a nupkg started with a number and the semver info helper broke :(

Also touches #522 around build information.

### CODE REVIEW

**Comment by baronfel** [comment]
Also addresses #1154

**Comment by forki** [comment]
do we need to do the same in paket?

**Comment by baronfel** [comment]
Yeah, but the problem there is compounded by the fact that there are two, slightly different implementations of SemVer there already split between the bootstrapper and Paket.Core.

Is it worth splitting this out somehow and/or paket file including the source file over there?

File: src/test/Test.FAKECore/ReleaseNotesSpecs.cs [hunk 1]
```diff
@@ -132,7 +132,7 @@ public class when_parsing_many_pre_release_versions
 #### 2.0.0-alpha001 - December 15 2013
 * A";
         static readonly ReleaseNotesHelper.ReleaseNotes expected =
-            ReleaseNotesHelper.ReleaseNotes.New("2.0.0", "2.0.0-rc007", null, new[] { "A" }.ToFSharpList());
+            ReleaseNotesHelper.ReleaseNotes.New("2.0.0", "2.0.0-rc2", null, new[] { "A" }.ToFSharpList());
```
Comment (forki) [inline src/test/Test.FAKECore/ReleaseNotesSpecs.cs]: why this change?

File: src/test/Test.FAKECore/SemVerHelperSpecs.cs [hunk 2]
```diff
@@ -96,9 +96,9 @@ public class when_comparing_semvers
             () => SemVerHelper.parse("1.0.0-alpha.1")
                 .ShouldBeLessThan(SemVerHelper.parse("1.0.0-alpha.beta"));
 
-        It should_assume_alpha_is_smaller_tha_beta =
+        It should_assume_that_longer_prereleases_are_greater =
```
Comment (forki) [inline src/test/Test.FAKECore/SemVerHelperSpecs.cs]: I think it's better to keep the old tests as well

**Comment by forki** [comment]
yes I think I tried to split it out into a separate file but didn't suceed

File: src/test/Test.FAKECore/SemVerHelperSpecs.cs (hunk 2 above)
Comment (baronfel) [inline src/test/Test.FAKECore/SemVerHelperSpecs.cs]: I kept the test data, but per the spec for these inputs 1.0.0-beta is larger, because a shorter prerelease tag is bigger than a longer prerelease tag.

File: src/test/Test.FAKECore/ReleaseNotesSpecs.cs (hunk 1 above)
Comment (baronfel) [inline src/test/Test.FAKECore/ReleaseNotesSpecs.cs]: I made this change because the comparison logic in the spec changed. rc007 is now less than rc2, so the expected output for the given release notes data had to change to take the improved comparison rules into account.

Comment (forki) [inline src/test/Test.FAKECore/ReleaseNotesSpecs.cs]: Oh that's breaking in fake. Not sure if we really want that.

Comment (baronfel) [inline src/test/Test.FAKECore/ReleaseNotesSpecs.cs]: Current logic is wrong, though.  Part of the reason I chafed working with nuget was because they implemented almost-sorta-but-not-really-semver, so none of the existing libraries worked.  Do we want to do the same?  Besides, if people depend on that sorting, they can either 

1. use a prerelease of the form `rc.{increment}`, where increment is a number and thus compared numerically, or
2. use a prerelease of the form `rc{padding}{number}`, and be consistent, so `rc007` compared to `rc002`,

Comment (forki) [inline src/test/Test.FAKECore/ReleaseNotesSpecs.cs]: The problem is usually the existing packages. But maybe this is less of a
problem now. It was a huge problem couple of years ago before we starter
with paket

**Comment by baronfel** [comment]
re: splitting this out - it's in a separate file with no non-framework dependencies, so we could just paket include it in paket.  This could be an issue, though, as paket exposes SemVerInfo publically so I think we'd get name collisions, and Fake's SemVerInfo wouldn't be the same as Paket's.

**Comment by baronfel** [comment]
I think we want to have a 'correct' implementation of SemVer in Fake and Paket for internal use for sure.  Then, especially in the nuget-related areas of Fake and Paket, if we need to have a stripped down version of Semver that's up to nuget spec but not SemVer proper, we can have mapping functions that assert the invariants.

**Comment by baronfel** [comment]
Alternatively, we could just use kolectiv's SemVer.net (FParsec parser + nice tests) in both places? https://github.com/xyncro/semver.net. That would handle the issue with sharing types around. But then we have even more back-compat issues.

**Comment by baronfel** [comment]
It would be very easy to leave the current implementation, mark it as obsolete, and include this new code as a 'SemVer2Helper' or something like that.  But Paket is the real sticking point.  Do you have any guidance here as to what you prefer?

**Comment by forki** [comment]
good question. I still don't know which packages would be affected

**Comment by baronfel** [comment]
hey, good news! Nuget client has started supporting SemVer 2.0 as of 3.5.0-rc1! https://emgarten.com/2016/08/09/semver-2-0-0-with-nuget-3-5-0-rc1/

So maybe this gives us some cover to move forward?

**Comment by forki** [comment]
ok let's try it.

**Comment by forki** [comment]
thanks!

**Comment by baronfel** [comment]
Want me to mirror these changes to Paket?

**Comment by forki** [comment]
yes. that would be awesome
